core/network/helpers/stream_read_buffer.hpp

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_NETWORK_HELPERS_STREAM_READ_BUFFER_HPP
+#define KAGOME_NETWORK_HELPERS_STREAM_READ_BUFFER_HPP
+
+#include "network/helpers/stream_proxy.hpp"
+
+namespace libp2p::connection {
+  struct StreamReadBuffer : StreamProxy,
+                            std::enable_shared_from_this<StreamReadBuffer> {
+    std::shared_ptr<std::vector<uint8_t>> buffer;
```

Comment:
Hm, why does it need to be a shared_ptr while the StreamReadBuffer itself is already an enable_shared_from_this?

Reply:
Read callbacks captures self as weak_ptr, and buffer as shared_ptr.  
Boost docs (async_read) say that "caller must guarantee that read buffer remains valid until the handler is called".